In [1]:
import sys #system specific module
#computer system or OS er sathe python script interface
import numpy as np #complex computation or complex matrix
# eg. circuit theory, power system, V=IR, FFT (fast fourier transform)
# eee er matlab
import pandas as pd #data analysis
# eg. sensor theke smart meter, voltage, current
import sklearn # scikit learn; CLASSICAL ML
# forecasting, fault detection
import torch #pytorch ; DL and AI
# ADVANCED DL (NN neural network) creation

print ("---System and Environment Verification---")
print (f"Python Version: {sys.version.split()[0]}")
print (f"NumPy Version: {np.__version__}")
print (f"Pandas Version: {pd.__version__}")
print (f"Scikit learn Version: {sklearn.__version__}")
print (f"PyTorch CUDA Acceleration Available : {torch.cuda.is_available()}")

---System and Environment Verification---
Python Version: 3.13.15
NumPy Version: 2.1.3
Pandas Version: 2.2.3
Scikit learn Version: 1.6.1
PyTorch CUDA Acceleration Available : False


**NumPy (vectorized operations), Pandas (tabular data manipulation), Matplotlib/Seaborn (visualization), scikit-learn (classical algorithms).**

Environment Verification: Confirms Python version and pre-installed ML stack in Colab. CUDA Available: False proves that GPU hardware acceleration is inactive.

Feature Matrix {X}: Columns tenure_months, monthly_charges, support_calls and contract_type represent input features describing customer behavior.

Target Matrix {Y}: Column churn is the binary label (1=churned, 0=retained). Formulating {Y} is the core step of Supervised Learning.



In [2]:
np.random.seed(42)
n_samples = 1000

data = {
    "tenure_months": np.random.randint(1, 72, n_samples),
    "monthly_charges": np.random.uniform(20.0, 120.0, n_samples),
    "support_calls": np.random.poisson(lam=2, size=n_samples),
    "contract_type": np.random.choice(["Month-to-month", "One year", "Two year"], size=n_samples)
}

df = pd.DataFrame(data)

churn_logit= -0.05 * df['tenure_months']+ 0.4 * df ['support_calls'] - 1.0
churn_prob= 1/ (1+np.exp(-churn_logit))
df['churn']=(churn_prob>np.random.uniform(0, 1, n_samples)).astype(int)

print("---Generated Dataset Sample(Problem Frame: Feature Matrix X + Target y)---")
print(df.head())

print ("---Class Balance Inspection (y distribution)---")
print (df['churn'].value_counts(normalize=True))

---Generated Dataset Sample(Problem Frame: Feature Matrix X + Target y)---
   tenure_months  monthly_charges  support_calls   contract_type  churn
0             52       105.569647              2        One year      0
1             15       103.021986              3  Month-to-month      0
2             61        59.718353              3  Month-to-month      0
3             21        86.808514              1  Month-to-month      1
4             24        40.498430              3  Month-to-month      0
---Class Balance Inspection (y distribution)---
churn
0    0.827
1    0.173
Name: proportion, dtype: float64


Class Imbalance (82.7% vs 17.3%): It shows why standard accuracy is misleading. If a dummy model predicts 0 for everyone, it achieves 82.7% accuracy while detecting zero churners. Therefore, we need F1 Score or ROC-AUC is necessary here.